lighthouse-core/test/lib/page-functions-test.js

```diff
@@ -0,0 +1,82 @@
+/**
+ * @license Copyright 2017 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const assert = require('assert');
+const fs = require('fs');
+const jsdom = require('jsdom');
+const URL = require( '../../lib/url-shim' );
+const DOM = require('../../report/html/renderer/dom.js');
+const Util = require('../../report/html/renderer/util.js');
+const pageFunctions = require('../../lib/page-functions');
+
+const TEMPLATE_FILE = fs.readFileSync(__dirname +
+    '/../../report/html/templates.html', 'utf8');
+
+/* eslint-env jest */
+
+describe('DetailsRenderer', () => {
+  let dom;
+
+  beforeAll(() => {
+    global.URL = URL;
+    global.Util = Util;
+    const document = jsdom.jsdom(TEMPLATE_FILE);
+    dom = new DOM(document);
+  });
+
+  afterAll(() => {
+    global.URL = undefined;
+    global.Util = undefined;
+  });
+
+  describe('get outer HTML snippets', () => {
+    it('gets full HTML snippet', () => {
+      const snippet = pageFunctions.getOuterHTMLSnippet(
+        dom.createElement('div', '', {id: '1', style: 'style'})
+      );
+      assert.ok(snippet.includes('div') &&
```

Comment:
hmm, it seems reasonable to assert the full string of `snippet` since the input element is fully under control of the test. It might make the test more readable as well, and could catch things like if someone messes up the regex at some point and it accidentally includes part of the closing tag or something.

WDYT vs asserting the parts like this is doing?

Reply:
Haha, yeah that's the way I did it originally, but was nervous that might be too explicit.  Changed it back!  Def better and will catch errors in parsing the carrots.  My only concern was with the ordering of elements in the string repr of the element, like style coming before aria-labels, but that is deterministic? maybe? Seems to work fine tho.

Reply:
> but that is deterministic

I think at least for jsdom we can assume it is. I have no idea if it's guaranteed in a real browser :)